[Ability] Fully implement Flower Gift and Victory Star

SirzBenjie opened the pull request:
<!-- (Once you have read these comments, you are free to remove them) -->
<!-- Feel free to look at other PRs for examples -->
<!--
Make sure the title includes categorization (choose the one that best fits):
-       [Bug]: If the PR is primarily a bug fix
-      [Move]: If a move has new or changed functionality
-   [Ability]: If an ability has new or changed functionality
-      [Item]: For new or modified items
-   [Mystery]: For new or modified Mystery Encounters
-      [Test]: If the PR is primarily adding or modifying tests
-     [UI/UX]: If the PR is changing UI/UX elements
-     [Audio]: If the PR is adding or changing music/sfx
-    [Sprite]: If the PR is adding or changing sprites
-   [Balance]: If the PR is related to game balance
- [Challenge]: If the PR is adding or modifying challenges
-  [Refactor]: If the PR is primarily rewriting existing code
-      [Docs]: If the PR is just adding or modifying documentation (such as tsdocs/code comments)
-    [GitHub]: For changes to GitHub workflows/templates/etc
-      [Misc]: If no other category fits the PR
-->
<!--
Make sure that this PR is not overlapping with someone else's work
Please try to keep the PR self-contained (and small)
-->

## What are the changes the user will see?
<!-- Summarize what are the changes from a user perspective on the application -->
Flower Gift and Victory Star will no longer be marked as partial. Flower gift will now boost the SpDef and Attack of allies during sun / harsh sun. Victory Star will now increase the accuracy of allies.

## Why am I making these changes?
<!--
Explain why you decided to introduce these changes
Does it come from an issue or another PR? Please link it
Explain why you believe this can enhance user experience
-->
<!--
If there are existing GitHub issues related to the PR that would be fixed,
you can add "Fixes #[issue number]" (ie: "Fixes #1234") to link an issue to your PR
so that it will automatically be closed when the PR is merged.
-->
Cherrim, what a great Pokemon. It deserves to have its ability fully implemented. So I did! Cherrim lovers will now get to see flower gift buffing allies as it should.
Since this required adding a class to increase stats for allies, it made sense to also fully implement Victini's victory star. So that's why this is in the same PR.

## What are the changes from a developer perspective?
<!--
Explicitly state what are the changes introduced by the PR
You can make use of a comparison between what was the state before and after your PR changes
Ex: What files have been changed? What classes/functions/variables/etc have been added or changed?
-->
The changes to get this working were slightly involved.
#### In `src/data/ability.ts`. 
Added a new class, `AllyStatMultiplierAbAttr` and a corresponding `applyFieldStatMultiplierAbAttrs` function.
The new class extends `AbAttr` and mirrors the logic of the existing `StatMultiplierAbAttr`. The class has an `ignorable` field instead of the `condition` field, as some ally stat multiplier abilities are ignorable and some are not. (The `condition` field from `StatMultiplierAbAttr` is related to accuracy / attack for physical / other moves.

In the flower gift ability attribute list, added two conditional attributes that check if the weather is sun/harsh sun, and add the `AllyStatMultiplierAbAttr`  attribute for the Attack/SpDef stat with a multiplier of 1.5. I also removed `.partial()` as Flower Gift is fully implemented with this change.

Similarly, for Victory Star, added the `AllyStatMultiplierAbAttr` attribute for accuracy with a multiplier of 1.1 and removed `.partial()`.

#### `src/data/pokemon.ts`

Added two parameters for ignoring the ally / target ally's ability to `getAttackDamage` and `getBaseDamage` that default to false.
I updated the callsites for these methods to properly provide the arguments - `getAttackDamage` passes along the arguments to its call to `getBaseDamage`; `getNextMove()` will conditionally ignore the target's ally ability if it has been revealed, and will never ignore its own ally's ability.

####  `src/data/move.ts`

In `ShellSideArmCategoryAttr`, added arguments to the call to `getBaseDamage` corresponding to the new parameters added to the function.


#### Tests

For `src/test/abilities/flower_gift.test.ts`: 
Uncommented the lines in the test for raising stats to verify the ally's stats are raised.
Added tests to ensure that damage is properly increased / reduced (or ignored) when attacking / getting hit by a special attack (respecting mold breaker and kin).

Created `src/test/abilities/victory_star.test.ts` with two tests: one that ensures the accuracy of the user is increased, and another to ensure the accuracy of the user's ally is increased.

## Screenshots/Videos
<!--
If your changes are changing anything on the user experience, please provide visual proofs of it
Please take screenshots/videos before and after your changes, to show what is brought by this PR
-->

https://github.com/user-attachments/assets/601309b6-1f05-4eb0-a248-e6d9823e5f13

https://github.com/user-attachments/assets/f99c5647-0abb-4232-8952-85d3397b31aa



## How to test the changes?
<!--
How can a reviewer test your changes once they check out on your branch?
Did you make use of the `src/overrides.ts` file?
Did you introduce any automated tests?
Do the reviewers need to do something special in order to test your changes?
-->

For the automed tests:
For flower gift : `npx vitest run src/test/abilities/flower_gift.test.ts` 
For victory star: `npx vitest run src/test/abilities/victory_star.test.ts`

For the manual tests to flower gift, in in `src/overrides.ts`, add overrides to have the starter mon be Cherrim. Change the moveset to be tackle, sunny day, mud slap, and recover. Add an override to make it a double battle. Change the enemy moveset to be splash, and the species of both mons to be bidoof.


In the first test, have cherrim use splash, and have the ally use tackle on the left bidoof.
Next turn, have cherrim use sunny day, and have the ally use tackle on the other bidoof. Observe that the second bidoof took ~50% more damage now that flower gift has activated.

For the second test, have cherrim use mud slap on the ally, and have the ally use splash. Move your mouse to the location on the ally's healthbar to mark how much damage it took.
Next turn, use sunny day with cherrim and recover with the ally.
Next turn, have cherrim use mud slap on the ally. Observe that ally's remaining health bar is gerater than the position the mouse was at, indicating that the SpDef has increased.

A manual test for victory star is not feasible due to it modifying accuracy.
## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test`)
  - [x] Have I created new automated tests (`npm run create-test`) or updated existing tests related to the PR's changes?
- [x] Have I provided screenshots/videos of the changes (if applicable)?
  - ~~[ ] Have I made sure that any UI change works for both UI themes (default and legacy)?~~

~~Are there any localization additions or changes? If so:~~
- ~~[ ] Has a locales PR been created on the [locales](https://github.com/pagefaultgames/pokerogue-locales) repo?~~
  - ~~[ ] If so, please leave a link to it here:~~ 
- ~~[ ] Has the translation team been contacted for proofreading/translation?~~

## Additional Comments

The reason this PR includes implementations for two abilities is that they both rely on the new `AllyStatMultiplierAbAttr` class being in the codebase. It didn't make sense to me to split this into 2 PRs because of that.

I considered whether a new class needed to be added for AllyStatMultiplerAbAttr. Originally, I had considered just extending `StatMultiplerAbAttr`. The issue with that approach would be twofold: First, abilities that *only* boosts the stats of the ally, and not the user itself. Second, the portions of code which look for `StatMultiplierAbAttr` match against the holder of the ability, (e.g. in `applyAbAttrsInternal`). This would make the logic a bit clunky for a small case.

The other thing I want to note is the decision to add parameters for ignoring / not ignoring the ability. This required adding two parameters to the `getBaseDamage` and `getAttackDamage` functions. I briefly considered hijacking the `ignoreAbility` / `ignoreSourceAbility` parameters for this. However, this is a bit misleading and couples ignoring the ability of the target / source with that of the ally. Explicit is better than implicit. While it is true that the only situation which currently uses these arguments is for Shell Side Arm, that implementation chose to split out ignoring ability and ignoring source ability. I followed suit here.

For these methods, there are now four arguments related to ignoring abilities. It may be worth considering a refactor that puts ability ignores into its own class, and use that instead of this pair of 4.

For the `getEffectiveStat`, I also had to add a parameter here for ignoring the ally ability. Just one though, since the pokemon only cares about its own ally, and not the target.
I did search for invocations of `getEffectiveStat` to update this, but found no uses of that argument. It's unclear why these arguments exist, but refactoring the code to remove them would be outside the scope of this PR.

DayKev commented on src/field/pokemon.ts:
```diff
@@ -964,6 +965,10 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
       applyStatMultiplierAbAttrs(StatMultiplierAbAttr, this, stat, statValue, simulated);
     }
+    const ally = this.getAlly();
+    if (!!ally) {
```
```suggestion
    if (ally) {
```

DayKev commented on src/field/pokemon.ts:
```diff
@@ -2570,9 +2575,17 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
     applyStatMultiplierAbAttrs(StatMultiplierAbAttr, target, Stat.EVA, evasionMultiplier);
 
+    const ally = this.getAlly();
+    if (!!ally) {
```
```suggestion
    if (ally) {
```

DayKev commented on src/field/pokemon.ts:
```diff
@@ -2570,9 +2575,17 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
 
     applyStatMultiplierAbAttrs(StatMultiplierAbAttr, this, Stat.ACC, accuracyMultiplier, false, sourceMove);
 
+
```
```suggestion
```

DayKev commented on src/data/ability.ts:
```diff
@@ -1581,10 +1581,58 @@ export class StatMultiplierAbAttr extends AbAttr {
   }
 }
 
+
```
```suggestion
```

DayKev commented on src/data/ability.ts:
```diff
@@ -1581,10 +1581,58 @@ export class StatMultiplierAbAttr extends AbAttr {
 
+    return false;
+  }
+
```
```suggestion
```

DayKev commented on src/data/ability.ts:
```diff
@@ -1581,10 +1581,58 @@ export class StatMultiplierAbAttr extends AbAttr {
+    this.stat = stat;
+    this.multiplier = multiplier;
+    this.ignorable = ignorable;
+
```
```suggestion
```

DayKev commented on src/data/ability.ts:
```diff
@@ -1581,10 +1581,58 @@ export class StatMultiplierAbAttr extends AbAttr {
+  }
+
+  /**
+   * applyAllyStat: Multiply a Pokemon's Stat due to an Ally's ability.
+   * @param pokemon {@linkcode Pokemon} the ally Pokemon with the ability (unused)
+   * @param passive {@linkcode boolean} unused
+   * @param simulated {@linkcode boolean} whether the ability is being simulated (unused)
+   * @param stat {@linkcode Stat} the type of the checked stat
+   * @param statValue {@linkcode Utils.NumberHolder} the value of the checked stat
+   * @param checkedPokemon {@linkcode Pokemon} the Pokemon this ability is targeting (unused)
+   * @param ignore {@linkcode boolean} Whether the ability should be ignored if possible
+   * @param args {any[]} unused
+   * @returns true if this changed the checked stat, false otherwise.
```
```suggestion
   * Multiply a Pokemon's Stat due to an Ally's ability.
   * @param pokemon - The ally {@linkcode Pokemon} with the ability (unused)
   * @param passive - unused
   * @param simulated - Whether the ability is being simulated (unused)
   * @param stat - The type of the checked {@linkcode Stat}
   * @param statValue - {@linkcode Utils.NumberHolder} containing the value of the checked stat
   * @param checkedPokemon - The {@linkcode Pokemon} this ability is targeting (unused)
   * @param ignoreAbility - Whether the ability should be ignored if possible
   * @param args - unused
   * @returns `true` if this changed the checked stat, `false` otherwise.
```

DayKev commented on src/data/ability.ts:
```diff
@@ -1581,10 +1581,58 @@ export class StatMultiplierAbAttr extends AbAttr {
+   * @param args {any[]} unused
+   * @returns true if this changed the checked stat, false otherwise.
+   */
+  applyAllyStat(pokemon: Pokemon, _passive: boolean, simulated: boolean, stat: BattleStat, statValue: Utils.NumberHolder, checkedPokemon: Pokemon, ignore: boolean, args: any[]): boolean | Promise<boolean> {
+    if (stat === this.stat && !(ignore && this.ignorable)) {
```
```suggestion
  applyAllyStat(pokemon: Pokemon, _passive: boolean, simulated: boolean, stat: BattleStat, statValue: Utils.NumberHolder, checkedPokemon: Pokemon, ignoreAbility: boolean, args: any[]): boolean | Promise<boolean> {
    if (stat === this.stat && !(ignoreAbility && this.ignorable)) {
```

DayKev commented on src/data/ability.ts:
```diff
@@ -5174,6 +5222,22 @@ export function applyStatMultiplierAbAttrs(attrType: Constructor<StatMultiplierA
+
+/**
+ * Applies an ally's Stat multiplier attribute
+ * @param attrType {@linkcode AllyStatMultiplierAbAttr} should always be AllyStatMultiplierAbAttr for the time being
+ * @param pokemon {@linkcode Pokemon} the pokemon with the ability
+ * @param stat {@linkcode Stat} the type of the checked stat
+ * @param statValue {@linkcode Utils.NumberHolder} the value of the checked stat
+ * @param checkedPokemon {@linkcode Pokemon} the Pokemon with the checked stat
+ * @param ignore {@linkcode boolean} Whether or not the ability should be ignored by the pokemon or its move.
+ * @param args unused
```
```suggestion
 * @param attrType - {@linkcode AllyStatMultiplierAbAttr} should always be AllyStatMultiplierAbAttr for the time being
 * @param pokemon - The {@linkcode Pokemon} with the ability
 * @param stat - The type of the checked {@linkcode Stat}
 * @param statValue - {@linkcode Utils.NumberHolder} containing the value of the checked stat
 * @param checkedPokemon - The {@linkcode Pokemon} with the checked stat
 * @param ignoreAbility - Whether or not the ability should be ignored by the pokemon or its move.
 * @param args - unused
```

DayKev commented on src/data/ability.ts:
```diff
@@ -5174,6 +5222,22 @@ export function applyStatMultiplierAbAttrs(attrType: Constructor<StatMultiplierA
+ * @param args unused
+ */
+export function applyAllyStatMultiplierAbAttrs(attrType: Constructor<AllyStatMultiplierAbAttr>,
+  pokemon: Pokemon, stat: BattleStat, statValue: Utils.NumberHolder, simulated: boolean = false, checkedPokemon, ignore, ...args: any[]): Promise<void> {
+  return applyAbAttrsInternal<AllyStatMultiplierAbAttr>(attrType, pokemon, (attr, passive) => attr.applyAllyStat(pokemon, passive, simulated, stat, statValue, checkedPokemon, ignore, args), args);
```
```suggestion
  pokemon: Pokemon, stat: BattleStat, statValue: Utils.NumberHolder, simulated: boolean = false, checkedPokemon: Pokemon, ignoreAbility: boolean, ...args: any[]
): Promise<void> {
  return applyAbAttrsInternal<AllyStatMultiplierAbAttr>(attrType, pokemon, (attr, passive) => attr.applyAllyStat(pokemon, passive, simulated, stat, statValue, checkedPokemon, ignoreAbility, args), args);
```

SirzBenjie commented:
Something strange is happening. I'm not quite sure why these changes are now causing `src/test/moves/order_up.test.ts > Moves - Order Up > should be boosted by Sheer Force while still applying a stat boost` to fail.

In debugging the test, I first wanted to see whether the sheer force damage boost was being applied, so I added a call to `vi.spyOn(orderUp, "calculateBattlePower")` to see, following the logic of similar tests in `src/test/abilities/sheer_force.test.ts`.
I changed the test to check if the move's power boost was expected (instead of whether or not `SHEER_FORCE` was in the list of abilities that were applied.
The test passes.

I'm entirely unsure how sheer force would avoid being added to `battleData.abilitiesApplied`, given I didn't change anything in `applyAbAttrsInternal`.

damocleas commented:
Going to need updates

DayKev reviewed (approved):
`getEffectiveStat` needs to be changed to `getEffectiveStat(stat, { ... })` but that's for a different PR.

SirzBenjie commented:
> `getEffectiveStat` needs to be changed to `getEffectiveStat(stat, { ... })` but that's for a different PR.

Yeah I agree.
